### PR TITLE
145 add support for constructing from data without copying (ktensor)

### PR DESCRIPTION
**Pull request by dmdunla**
Addresses the `ktensor` part of:
- #145 
- 

----
:books: Documentation preview :books:: https://pyttb--182.org.readthedocs.build/en/182/

### CODE REVIEW

**Comment by dmdunla** [comment]
The following methods currently return a copy make a copy of the ktensor:
- `permute`
- `symmetrize`

The following methods alter the underlying data members and do not make a copy:
- `arrange`
- `fixsigns`
- `normalize`

Is this what we want? Which should be in-place (no copy) and which should make copies?

**Comment by dmdunla** [comment]
@ntjohnson1 Please review changes to `ktensor` at this point. Once this has been reviewed, I'll move to tensor.

**Comment by ntjohnson1** [comment]
> The following methods currently return a copy make a copy of the ktensor:
> 
> * `permute`
> * `symmetrize`
> 
> The following methods alter the underlying data members and do not make a copy:
> 
> * `arrange`
> * `fixsigns`
> * `normalize`
> 
> Is this what we want? Which should be in-place (no copy) and which should make copies?

We originally decided to match MATLAB's in-place vs copy choice. We are consistent with the methods you listed.

Returns copy
* [`permute`](https://gitlab.com/tensors/tensor_toolbox/-/blob/dev/%40ktensor/permute.m)
* [`symmetrize`](https://gitlab.com/tensors/tensor_toolbox/-/blob/dev/%40ktensor/symmetrize.m)

In-place
* [`arrange`](https://gitlab.com/tensors/tensor_toolbox/-/blob/dev/%40ktensor/arrange.m)
* [`fixsigns`](https://gitlab.com/tensors/tensor_toolbox/-/blob/dev/%40ktensor/fixsigns.m)
* [`normalize`](https://gitlab.com/tensors/tensor_toolbox/-/blob/dev/%40ktensor/normalize.m)

**Review by ntjohnson1 — COMMENTED** [review]
Looks good to me. A few NITs you can address at your discretion. Always a fan of smaller PRs, my vote would be to merge this then open a second PR for tensor since the changes should be aligned in spirit but I doubt the implementation overlaps.

File: pyttb/ktensor.py [hunk 1]
```diff
@@ -100,186 +96,74 @@ def from_data(cls, weights, *factor_matrices):
...
-         [7. 8.]]
+        # Empty constructor
+        if factor_matrices is None and weights is None:
+            self.weights = np.array([])
+            self.factor_matrices = []
+            return
+
+        # 'factor_matrices' must be a list
+        if not isinstance(factor_matrices, list):
+            assert False, "Input 'factor_matrices' must be a list."
```
[the remark this answers is not in the document]
Comment (dmdunla) [reply pyttb/ktensor.py:129]: OK, will leave this to another issue/PR

File: pyttb/ktensor.py [hunk 2]
```diff
@@ -2412,13 +2296,18 @@ def __repr__(self):
         """
         s = ""
         s += "ktensor of shape "
-        s += (" x ").join([str(int(d)) for d in self.shape])
+        s += "("
+        s += (", ").join([str(int(d)) for d in self.shape])
+        s += ")"
```
[the remark this answers is not in the document]
Comment (dmdunla) [reply pyttb/ktensor.py:2301]: Yes, I had not discussed this change before making it. We had not been testing output of empty ktensors, so I opted to move away from the MATLAB style towards a tuple-based, numpy-like style for __repr__. I'll add an issue to discuss this and change back if we decide that.

File: pyttb/tensor.py [hunk 3]
```diff
@@ -671,7 +671,10 @@ def mttkrp(self, U: Union[ttb.ktensor, List[np.ndarray]], n: int) -> np.ndarray:
 
         # extract the list of factor matrices if given a ktensor
         if isinstance(U, ttb.ktensor):
-            U = ttb.ktensor.from_tensor_type(U)
+            # TODO: the following causes an error in typing,
+            # but I think this is what we want
+            # U = U.copy()
```
[the remark this answers is not in the document]
Comment (dmdunla) [reply pyttb/tensor.py:676]: Added type hint.

File: tests/test_ktensor.py [hunk 4]
```diff
@@ -56,88 +56,80 @@ def sample_ktensor_symmetric():
...
-    # won't work with instances other than ktensors
+    # Input: factor_matrices provided, weights = None
+    K = ttb.ktensor(data["factor_matrices"])
+    assert (
+        K.weights
+        == np.ones(
+            2,
+        )
+    ).all()
+    assert (K.factor_matrices[0] == data["factor_matrices"][0]).all()
```
[the remark this answers is not in the document]
Comment (dmdunla) [reply tests/test_ktensor.py:86]: Will add new issue for this.

**Comment by dmdunla** [comment]
Will finalize this PR to address `ktensor` only in #145.

@ntjohnson1 Would you suggest breaking up #145 or just submitting several PRs for the individual classes?

**Comment by ntjohnson1** [comment]
> Will finalize this PR to address `ktensor` only in #145.
> 
> @ntjohnson1 Would you suggest breaking up #145 or just submitting several PRs for the individual classes?

Whatever you think is most obvious. No strong preference. I mostly just wanted this to land today since I branched off it to start in on ktensor linting and typing [here](https://github.com/ntjohnson1/pyttb/tree/ktensor_lint_type) (finished linting and halfway through typing).

**Review by ntjohnson1 — APPROVED** [review]

**Pull request merged**